src/version_set.cc

```diff
@@ -204,52 +207,61 @@ Status VersionSet::LogAndApply(VersionEdit* edit) {
-    if (blob_it == files.end()) {
-      fprintf(stderr, "blob file %" PRIu64 " doesn't exist before\n", number);
-      abort();
-    } else if (blob_it->second->is_obsolete()) {
-      fprintf(stderr, "blob file %" PRIu64 " has been deleted before\n",
-              number);
-      abort();
+Status VersionSet::Apply(EditCollector* collector) {
+  for (auto& cf : collector->column_families_) {
+    auto cf_id = cf.first;
+    auto it = column_families_.find(cf_id);
+    if (it == column_families_.end()) {
+      // TODO: support OpenForReadOnly which doesn't open DB with all column
+      // family so there are maybe some invalid column family, but we can't just
+      // skip it otherwise blob files of the non-open column families will be
+      // regarded as obsolete and deleted.
+      continue;
+    }
+    auto& files = it->second->files_;
+
+    for (auto& file : cf.second.deleted_files_) {
+      auto number = file.first;
+      auto blob_it = files.find(number);
+      if (blob_it == files.end()) {
+        ROCKS_LOG_ERROR(db_options_.info_log,
+                        "blob file %" PRIu64 " doesn't exist before\n", number);
+        return Status::Corruption("blob file doesn't exist before");
+      } else if (blob_it->second->is_obsolete()) {
+        ROCKS_LOG_ERROR(db_options_.info_log,
+                        "blob file %" PRIu64 " has been deleted already\n",
+                        number);
+        return Status::Corruption("blob file has been deleted already");
+      }
+      it->second->MarkFileObsolete(blob_it->second, file.second);
     }
-    it->second->MarkFileObsolete(blob_it->second, file.second);
-  }
 
-  for (auto& file : edit->added_files_) {
-    auto number = file->file_number();
-    auto blob_it = files.find(number);
-    if (blob_it != files.end()) {
-      if (blob_it->second->is_obsolete()) {
-        fprintf(stderr, "blob file %" PRIu64 " has been deleted before\n",
-                number);
-      } else {
-        fprintf(stderr, "blob file %" PRIu64 " has been added before\n",
-                number);
+    for (auto& file : cf.second.added_files_) {
+      auto number = file.first;
+      auto blob_it = files.find(number);
+      if (blob_it != files.end()) {
+        if (blob_it->second->is_obsolete()) {
+          ROCKS_LOG_ERROR(db_options_.info_log,
+                          "blob file %" PRIu64 " has been deleted before\n",
+                          number);
+          return Status::Corruption("blob file has been deleted before");
+        } else {
+          ROCKS_LOG_ERROR(db_options_.info_log,
+                          "blob file %" PRIu64 " has been added before\n",
+                          number);
+          return Status::Corruption("blob file has been added before");
+        }
       }
-      abort();
+      it->second->AddBlobFile(file.second);
     }
-    it->second->AddBlobFile(file);
-  }
 
-  it->second->ComputeGCScore();
+    it->second->ComputeGCScore();
+  }
```

Comment:
let's also sanity check those files deleted after added in `EditCollector` does not exist in `VersionSet`.